### DISCUSSION TITLE
Performance of FMU implementation

### DISCUSSION THREAD

**Discussion by FabiJa** [casadi/casadi#3591] [Q&A]
Hi,

I am wondering, what the bottlenecks are considering the performance degradation when comparing a model encapsulated in a FMU and a direct implementation of the model in CasADi?
In the publication "Parameter Estimation of Modelica Building Models Using CasADi" a performance degradation of factor 20 is observed. I did some comparison for the `bvsm.mo` example provided in a CasADi tutorial at the last Modelica conference and here it was only a factor of 2-3. On the other hand, I tested another simple model (ODE with 1 state and some intermedate quantities but no algebraic loop, i.e. comparable complexity as `bvsm.mo` ) leading to a performance degradation of factor > 100.

So in summary, I am wondering if the performance degradation is due to the FMI interface itself ( see e.g. [here](https://github.com/modelica/fmi-standard/issues/1902)) or if something can be done on the CasADi implementation?
Are there in general some advices to optimize the performance when using CasADi and a FMU?

Thanks.

**Comment by jaeandersson** [comment]
As of right now, the FMI interface in CasADi is really an advanced prototype and essentially unsupported if something goes wrong. In addition to general code immaturity, one fundamental limitation is that it currently doesn't support FMI 3, which means that the adjoint derivatives are calculated quite inefficiently. In general, you will probably always see a large overhead for very small dimensions (such as the `bvsm.mo` example), but for large-scale, that overhead shouldn't really matter. Being able to calculate multiple derivatives at once (the issue you referenced) would be nice and could give a significant speedup, but it's also not crucial - there are plenty of other ways to get efficiency improvements.

Unfortunately. as of now, the FMI interface is is in a bit of a standstill pending clarity of the funding situation. Hopefully that will be clarified in the next few weeks. There is a chance that the FMI interface will be removed from CasADi and moved into a separate library with a more restrictive licensing for industrial use. Otherwise, I don't see how to justify continuing the development. In any case, I would like to find a way to get the FMI interface mature and more efficient.
